File: epi_judge_java/epi/SpiralOrdering.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.List;

public class SpiralOrdering {

    @EpiTest(testDataFile = "spiral_ordering.tsv")
    public static List<Integer> matrixInSpiralOrder(List<List<Integer>> squareMatrix) {
        if (squareMatrix.isEmpty()) return List.of();
        int n = squareMatrix.size();
        List<Integer> ans = new ArrayList<>();
        int right = n - 1;
        int left = 0;
        int top = 0;
        int bottom = n - 1;
        while (ans.size() < n * n) {
            for (int i = left; i <= right; i++) {
                ans.add(squareMatrix.get(top).get(i));
            }
            top++;
            for (int i = top; i <= bottom; i++) {
                ans.add(squareMatrix.get(i).get(right));
            }
            right--;
            for (int i = right; i >= left; i--) {
                ans.add(squareMatrix.get(bottom).get(i));
            }
            bottom--;
            for (int i = bottom; i >= top; i--) {
                ans.add(squareMatrix.get(i).get(left));
            }
            left++;
        }
        return ans;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "SpiralOrdering.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/spiral_ordering.cc
#include <vector>

#include "test_framework/generic_test.h"

using std::vector;

vector<int> MatrixInSpiralOrder(const vector<vector<int>> &square_matrix) {
    vector<int> result;
    if (square_matrix.empty() || square_matrix[0].empty()) return result;

    int rows = square_matrix.size();
    int columns = square_matrix[0].size();

    int top = 0;
    int bottom = rows - 1;
    int left = 0;
    int right = columns - 1;

    while (result.size() < rows * columns) {
        for (int c = left; c <= right; ++c) {
            result.push_back(square_matrix[top][c]);
        }
        ++top;

        for (int r = top; r <= bottom; ++r) {
            result.push_back(square_matrix[r][right]);
        }
        --right;

        for (int c = right; c >= left; --c) {
            result.push_back(square_matrix[bottom][c]);
        }
        --bottom;

        for (int r = bottom; r >= top; --r) {
            result.push_back(square_matrix[r][left]);
        }
        ++left;
    }

    return result;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"square_matrix"};
    return GenericTestMain(args, "spiral_ordering.cc", "spiral_ordering.tsv",
                           &MatrixInSpiralOrder, DefaultComparator{},
                           param_names);
}
